compose/bundle.py

```diff
@@ -108,27 +108,24 @@ def get_image_digest(service, allow_fetch=False):
         # digests
         return image['RepoDigests'][0]
 
-    if not allow_fetch:
-        if 'build' in service.options:
-            raise NeedsPush(service.image_name)
-        else:
-            raise NeedsPull(service.image_name)
+    if 'build' not in service.options:
+        raise NeedsPull(service.image_name)
 
-    return fetch_image_digest(service)
+    if not allow_push:
+        raise NeedsPush(service.image_name)
 
+    return push_image(service)
 
-def fetch_image_digest(service):
-    if 'build' not in service.options:
-        digest = service.pull()
-    else:
-        try:
-            digest = service.push()
-        except:
-            log.error(
-                "Failed to push image for service '{s.name}'. Please use an "
-                "image tag that can be pushed to a Docker "
-                "registry.".format(s=service))
-            raise
+
+def push_image(service):
+    try:
+        digest = service.push()
+    except:
+        log.error(
+            "Failed to push image for service '{s.name}'. Please use an "
+            "image tag that can be pushed to a Docker "
+            "registry.".format(s=service))
```

Comment:
I think it would be good to include the original error message here as well. The push could fail with an auth error, which would make this message misleading, but at least if we include the original error message they should have an idea of how to fix it.

It could also fail with some temporary connection error.